superset-frontend/spec/javascripts/explore/components/VizTypeControl_spec.jsx

```diff
@@ -31,6 +33,9 @@ const defaultProps = {
   isModalOpenInit: true,
 };
 
+const waitForEffects = () =>
+  act(() => new Promise(resolve => setTimeout(resolve, 0)));
```

Comment:
nifty! This seem like a better way to do this in rtl when there are issues of components not being ready.

Reply:
Actually, the recommended way here is to use `findBy` and `awaitFor` to deal with these situations. The `render` of RTL is [already wrapped](https://javascript.plainenglish.io/you-probably-dont-need-act-in-your-react-tests-2a0bcd2ad65c) in `act`.

Another point is to try to avoid [nesting](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing) in your tests.

Reply:
The issue here wasn't that components under test weren't ready. AntD and/or the Icon component seems to be doing some kind of async changes, so even though the test passes, there is a warning `an update to Icon was not wrapped  in act()`. This sufficiently `act`-ifies whatever side effects are going on and prevents those warnings.

Reply:
As for nesting, I generally agree with KCD but I also don't think the code here steps into any of the pitfalls in his article, and there's definitely a convention in this codebase of using `beforeEach` in this way.

Reply:
> The issue here wasn't that components under test weren't ready. AntD and/or the Icon component seems to be doing some kind of async changes, so even though the test passes, there is a warning `an update to Icon was not wrapped in act()`. This sufficiently `act`-ifies whatever side effects are going on and prevents those warnings. I'll add a comment in the code explaining this.

Here is a version using `awaitFor` and without warnings:

```
class MainPreset extends Preset {
  constructor() {
    super({
      name: 'Legacy charts',
      plugins: [
        new LineChartPlugin().configure({ key: 'line' }),
        new EchartsTimeseriesChartPlugin().configure({
          key: 'echarts_timeseries',
        }),
        new TimeTableChartPlugin().configure({ key: 'time_table' }),
        new EchartsMixedTimeseriesChartPlugin().configure({
          key: 'mixed_timeseries',
        }),
      ],
    });
  }
}

const getTestId = testWithId<string>(VIZ_TYPE_CONTROL_TEST_ID, true);

const defaultProps: VizTypeControlProps = {
  description: '',
  label: '',
  name: '',
  value: '',
  labelType: 'primary',
  onChange: jest.fn(),
  isModalOpenInit: true,
};

beforeAll(() => {
  new MainPreset().register();
});

const renderWrapper = async (
  props = defaultProps,
  state: object = stateWithoutNativeFilters,
) => {
  render(
    <DynamicPluginProvider>
      <VizTypeControl {...props} />
    </DynamicPluginProvider>,
    {
      useRedux: true,
      initialState: state,
    },
  );
};

test('Search visualization type', async () => {
  renderWrapper();

  const visualizations = screen.getByTestId(getTestId('viz-row'));

  expect(visualizations).toHaveTextContent(/Time-series Table/);
  expect(visualizations).toHaveTextContent(/Time-series Chart/);
  expect(visualizations).toHaveTextContent(/Mixed timeseries chart/);
  expect(visualizations).toHaveTextContent(/Line Chart/);

  // search
  await waitFor(() => {
    userEvent.type(
      screen.getByTestId(getTestId('search-input')),
      'time series',
    );
  });

  expect(visualizations).toHaveTextContent(/Time-series Table/);
  expect(visualizations).toHaveTextContent(/Time-series Chart/);
  expect(visualizations).toHaveTextContent(/Mixed timeseries chart/);
  expect(visualizations).not.toHaveTextContent(/Line Chart/);
});
```

Reply:
> As for nesting, I generally agree with KCD but I also don't think the code here steps into any of the pitfalls in his article, and there's definitely a convention in this codebase of using `beforeEach` in this way.

The `describe` was heavily used by Enzyme tests. When we migrated to RTL tests the `describe` was abandoned following KCD best practices. If you look for RTL files (.test) you'll see that the majority is not using `describe` anymore.

Reply:
These are only suggestions to make the code look more similar to other RTL tests. Whether you choose to apply them or not, this PR can be merged 😉